pkg/stash/stasher.go

```diff
@@ -0,0 +1,62 @@
+package stash
+
+import (
+	"context"
+	"time"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/module"
+	"github.com/gomods/athens/pkg/storage"
+)
+
+// Stasher has the job of taking a module
+// from an upstream entity and stashing it to a Storage Backend.
+type Stasher interface {
+	Stash(string, string) error
+}
+
+// Wrapper helps extend the main stasher's functionality with addons.
+type Wrapper func(Stasher) Stasher
+
+// New returns a plain stasher that takes
+// a module from a download.Protocol and
+// stashes it into a backend.Storage.
+func New(f module.Fetcher, s storage.Backend, wrappers ...Wrapper) Stasher {
+	var st Stasher = &stasher{f, s}
+	for _, w := range wrappers {
+		st = w(st)
+	}
+
+	return st
+}
+
+type stasher struct {
+	f module.Fetcher
+	s storage.Backend
+}
+
+func (s *stasher) Stash(mod, ver string) error {
+	const op errors.Op = "stasher.Stash"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*10)
+	defer cancel()
+	v, err := s.fetchModule(ctx, mod, ver)
```

Comment:
can we check if mod-ver is present.
what if call to singleflight.Stash comes for m,v at the time when it is already fetched and subscribers are notified.
after subs are notified, lock is released and stash continue with m,v with fetching because it finds out that nobody is working on m,v and it reported a cache miss
maybe

Reply:
@michalpristas that shouldn't happen (AFAIK). Because any request that comes in will *either* be part of the singleflight or will be gotten from storage when the download.Protocol checks. 
I can write a test (in another PR) that ensures a Stash is never hit more than once.